Argument validation

Sutharsan opened the pull request:
To prevent invalid API requests, validation of required arguments is added to 'country by uuid' and 'city by uuid' requests.

Sutharsan committed:
WIP

Sutharsan committed:
Added validation to country and city requests.
- Check added for required uuid and language codes.
- Validation added to uuid and multilingual traits.
- Tests added.

bartfeenstra commented:
The traits don't know about their usage, e.g. whether or not their functionality is required for the requests. As such, the exception messages and method names should not contain these assumptions. Instead they should say the values are invalid (and contain more specific information when available).

`// No assertion. Exception is thrown.`: no problem, but PHPUnit will automatically complain if a test method does not test anything (has no asserts, and does not expect exceptions, and sets no expectations).

`$this->sut = CityByUuid::create($this->productionRequestHandler);`
Don't use the production handler when simply testing validation. Instead, use the mocked handler in `RequestBaseTestBase::requestHandler`.

Sutharsan committed:
Modified exception messages.

Sutharsan committed:
Reworked argument validation tests.
- Exception tests must fail.
- Use DataProvider where possible.
- Removed duplicate tests.

Sutharsan committed:
Use mock request handler.

bartfeenstra commented on tests/Request/CityByUuidTest.php:
```diff
@@ -106,6 +106,27 @@ public function testExecuteRealRequest($form, $instanceof)
+          ->setLanguageCodes($language_codes)
+          ->execute();
+
+        $this->fail('Invalid arguments must throw an exception.');
```
This is redundant if `@expectedException` is specified. PHPUnit will log a test failure automatically if no exception of the specified type is thrown during the test.

Sutharsan replied:
Now removed. Your comment above "but PHPUnit will automatically complain ..." got me confused.

bartfeenstra commented on tests/Request/MultilingualTraitTest.php:
```diff
@@ -0,0 +1,101 @@
+     *
+     * @var string
+     */
+    protected $invalidLanguageCodes;
```
The property name and its docblock description are plural, yet the type hint indicates a single string value.

bartfeenstra commented on tests/Request/MultilingualTraitTest.php:
```diff
@@ -0,0 +1,101 @@
+        $reflection_method->setAccessible(true);
+        $reflection_method->invoke($this->sut);
+
+        $this->fail('Invalid language codes must throw an exception.');
```
Not necessary here either.

bartfeenstra commented on tests/Request/UuidTraitTest.php:
```diff
@@ -0,0 +1,100 @@
+
+        $this->sut->setUuid($this->invalidUuid);
+
+        $this->fail('Invalid UUID must throw an exception.');
```
Not necessary here either.

bartfeenstra commented on tests/Request/UuidTraitTest.php:
```diff
@@ -0,0 +1,100 @@
+        $reflection_method->setAccessible(true);
+        $reflection_method->invoke($this->sut);
+
+        $this->fail('Invalid UUID must throw an exception.');
```
Not necessary here either.

Sutharsan committed:
Remove redundant fail + message.

Sutharsan committed:
Fix type hint.
